### PR TITLE
Add batch fuse 

### PR DESCRIPTION
**Pull request by HuanyuCai**
Batch fuse is added by this pull request
If you want to use tflite-vx-delegate to run a tflite model to test batch fuse, you can use this updated but unofficial code [delegate_main.cc](https://github.com/HuanyuCai/tflite-vx-delegate/blob/batch_fuse/delegate_main.cc)

### CODE REVIEW

File: src/tim/transform/ops/pad_layout_inference.h [hunk 1]
```diff
@@ -44,6 +44,11 @@ class PadLayoutInfer : public OpLayoutInfer {
     auto i_src = op_->impl()->InputsTensor()[0];
     auto input_pv = context_->GetPermuteVector(i_src);
 
+    if (i_src->GetSpec().attr_ & tim::vx::TensorAttribute::INPUT){
+      input_pv = std::make_shared<PermuteVector<4>>(kCWHN2WHCN);
+      context_->UpdateTensorMap(i_src, InsertPermute(context_->GetMapedTensor(i_src), input_pv));
+    }
```
Comment (sunshinemyson) [inline src/tim/transform/ops/pad_layout_inference.h:50]: move this into standalone commit.

File: src/tim/vx/tensor.cc [hunk 2]
```diff
@@ -388,7 +388,7 @@ TensorSpec& TensorSpec::SetQuantization(Quantization& quantization) {
 }
 
 TensorSpec TensorSpec::AsTransientSpec() const {
-  return TensorSpec(this->datatype_, ShapeType({}), TensorAttribute::TRANSIENT,
+  return TensorSpec(this->datatype_, this->shape_, TensorAttribute::TRANSIENT,
```
Comment (sunshinemyson) [inline src/tim/vx/tensor.cc:391]: Original idea for Transient Tensor is : we suppose ovxlib will put correct shape after op_setup. Why we need to give the shape explicitly here?

File: src/tim/fuse/batch_fuse.cc [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+}
+
+bool BatchFuseContext::IsVisited(
+    const std::shared_ptr<vx::Operation>& op) const {
+  if (visited_op_.end() !=
+      std::find(visited_op_.begin(), visited_op_.end(), op)) {
+    return true;
+  } else {
+    return false;
+  }
```
Comment (sunshinemyson) [inline src/tim/fuse/batch_fuse.cc:48]: return (visited_op_.end() != std::find( ...)) ; instead of if-else.

File: src/tim/vx/tensor.cc (hunk 2 above)
Comment (HuanyuCai) [reply src/tim/vx/tensor.cc:391]: Before op_setup, batchfuse in timx-vx need tensor shape to compute the fused shape. if we do not set this shape explicitly, we will get empty shape when we run model in tflite-vx-delegate

File: src/tim/transform/ops/pad_layout_inference.h (hunk 1 above)
Comment (HuanyuCai) [reply src/tim/transform/ops/pad_layout_inference.h:50]: Done

File: src/tim/fuse/batch_fuse.cc (hunk 3 above)
Comment (HuanyuCai) [reply src/tim/fuse/batch_fuse.cc]: Done

File: src/tim/fuse/ops/reduce_batch_fuse.h [hunk 4]
```diff
@@ -0,0 +1,126 @@
+/****************************************************************************
+ *
+ *    Copyright (c) 2020 Vivante Corporation
```
Comment (sunshinemyson) [inline src/tim/fuse/ops/reduce_batch_fuse.h:3]: Copyright  2023

File: src/tim/fuse/ops/reduce_batch_fuse.h [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    // auto batch_src_axis = context_->GetBatchAxis();  // 3
+    auto fuse_src_axes = context_->GetFuseAxes();    // [1, 2]
+
+    auto perm_axis_map = context_->GetPermAxisMap(input_tensor);
+    auto fuse_axes = context_->GetPermFuseAxes(input_tensor);
+    auto batch_axis = context_->GetPermBatchAxis(input_tensor);
+    // auto c_axis = context_->GetPermChannelAxis(input_tensor);
+
+    // auto w_axis = fuse_axes[0];
+    // auto h_axis = fuse_axes[1];
```
Comment (sunshinemyson) [inline src/tim/fuse/ops/reduce_batch_fuse.h:83]: remove unused code

File: src/tim/fuse/ops/transpose_batch_fuse.h [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+  bool GapBackwardInference(
+      std::vector<std::shared_ptr<vx::Tensor>>& former_tensors) override {
+    auto input_tensor = op_->impl()->InputsTensor()[0];
+    //To hack a werror
+    former_tensors.push_back(input_tensor);
+    former_tensors.pop_back();
+    return false;
+  }
+
+#define CREATE_AND_CONCAT_OP(idx, start, length)                             \
```
Comment (sunshinemyson) [inline src/tim/fuse/ops/transpose_batch_fuse.h:72]: why define this here?

File: src/tim/fuse/batch_fuse_context.h [hunk 7]
```diff
@@ -0,0 +1,161 @@
+#ifndef TIM_VX_BATCH_FUSE_CONTEXT_H_
+#define TIM_VX_BATCH_FUSE_CONTEXT_H_
+// #include "permute_vector.h"
```
[the remark this answers is not in the document]
Comment (HuanyuCai) [reply src/tim/fuse/batch_fuse_context.h]: Done

File: src/tim/fuse/batch_fuse.cc [hunk 8]
```diff
@@ -0,0 +1,520 @@
+
+#include "tim/fuse/batch_fuse.h"
+// #include "builtin_op_impl.h"
```
[the remark this answers is not in the document]
Comment (HuanyuCai) [reply src/tim/fuse/batch_fuse.cc]: Done

File: src/tim/fuse/batch_fuse.cc [hunk 9]
```diff
@@ -0,0 +1,520 @@
...
+#include <algorithm>
+#include <deque>
+#include <cassert>
+
+#include "tim/vx/context.h"
+#include "tim/vx/graph.h"
+#include "tim/vx/operation.h"
+
+#include "ops/conv2d_batch_fuse.h"
+// #include "ops/pad_v2_batch_fuse.h"
```
[the remark this answers is not in the document]
Comment (HuanyuCai) [reply src/tim/fuse/batch_fuse.cc]: Done

File: src/tim/fuse/batch_fuse.cc [hunk 10]
```diff
@@ -0,0 +1,520 @@
...
+    tensor_queue.push_back(const_in);
+    tensor_queue_pad_forward.push_back(const_in);
+  }
+
+  while (!tensor_queue_pad_forward.empty()) {
+    //bfs
+    auto tensor = tensor_queue_pad_forward.front();
+    tensor_queue_pad_forward.pop_front();
+
+    //dfs
```
[the remark this answers is not in the document]
Comment (HuanyuCai) [reply src/tim/fuse/batch_fuse.cc]: Done

File: src/tim/fuse/ops/reduce_batch_fuse.h (hunk 5 above)
Comment (HuanyuCai) [reply src/tim/fuse/ops/reduce_batch_fuse.h]: Done

File: src/tim/fuse/ops/transpose_batch_fuse.h (hunk 6 above)
Comment (HuanyuCai) [reply src/tim/fuse/ops/transpose_batch_fuse.h]: Deleted

File: src/tim/fuse/batch_fuse_context.h [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    fuse_axes_ = fuse_axes;
+  }
+
+
+  // After graph layout infered, we get a layout infered graph
+  // Then we clone a new graph from layout infered graph, when we clone,
+  // we just clone ops and set new shape for tensors in cloned graph, here we just multiply fake_batch_ times of batch
+  // Lastly, we batch fuse the cloned graph to get a batch fused graph
+
+  const std::shared_ptr<vx::Graph>& src_graph_;   //layout infered graph
```
[the remark this answers is not in the document]
Comment (HuanyuCai) [reply src/tim/fuse/batch_fuse_context.h:121]: I'm not so sure either

File: src/tim/fuse/batch_fuse.cc [hunk 12]
```diff
@@ -0,0 +1,520 @@
...
+    return it->second;
+  } else {
+    VSILOGE("Tensor has not beed inserted in permute axis map.");
+    assert(false);
+  }
+
+  return {};
+}
+
+#define REGIST_BATCH_FUSE(op_idx, name)                              \
```
[the remark this answers is not in the document]
Comment (HuanyuCai) [reply src/tim/fuse/batch_fuse.cc]: Done

File: src/tim/fuse/batch_fuse.cc [hunk 13]
```diff
@@ -0,0 +1,520 @@
...
+  }
+  return std::make_pair(next_tensors, need_backward);
+}
+
+std::vector<std::shared_ptr<vx::Tensor>> HandleCloneGraph(
+    std::shared_ptr<batch_fuse_impl::BatchFuseContext>& ctx,
+    const std::shared_ptr<vx::Operation>& op) {
+  auto op_id = op->impl()->kind_;
+  std::vector<std::shared_ptr<vx::Tensor>> next_tensors;
+  switch (op_id) {
```
[the remark this answers is not in the document]
Comment (HuanyuCai) [reply src/tim/fuse/batch_fuse.cc]: Done

File: src/tim/fuse/batch_fuse_context.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (HuanyuCai) [reply src/tim/fuse/batch_fuse_context.h:125]: Has set to private
